content/blog/2022-02-22-profiling.md

```diff
@@ -0,0 +1,208 @@
+
+The naive way to count how many times each path executes is to exhaust every path in the CFG and label each one. However, this quickly becomes intractable as the CFG becomes large, or the CFG contains cycles with infinite paths. 
+
+In short, the motivation is to use an efficient path profiling method instead of traditional block/edge profiling to improve the accuracy without increasing too much effort. To have a preview, the overhead of path profiling is 30.9%, while the overhead of edge profiling is 16.1%. (You can decide whether this overhead is acceptable for you!)
+
+
+### Overview
+
+The essential idea behind path profiling is to encode paths with states. When passing through edges, we increment a register and naturally arrive at a unique number for each path. Furthermore, we can reduce the times we increment the register by placing “instrumentations” on some edges. This process is also “reversible”: we can find out the exact path in the Control-Flow Graph (CFG) given the encoding of that path. 
+
+The paper starts from a CFG with a single back edge and generalizes to CFGs with multiple back edges. The experiment section evaluates the efficient path profiling algorithm across SPEC95 benchmarks and demonstrates a 30.9% runtime overhead on average.
+
+
+### Generating Path Encodings
+
+How do we generate unique encoding for each path? We use the following simplified example to illustrate intuition.
+
+<p align="center">
+<img src="https://res.cloudinary.com/dxzx2bxch/image/upload/v1646251186/cs6120/Screen_Shot_2022-03-02_at_14.59.36_cpnltz.png" alt="alt_text" title="image_tooltip" style="zoom:20%;" />
+</p>
+
+<p align="center">
+<strong>A simplified example of generating path encoding  </strong>
+</p>
+
+There are four paths in the graph; we now assign a value to each edge to make the path sum unique. We label the first outgoing edge to zero at the root node. Since we know along that edge there are three paths, we label the second outgoing edge with three.  
+
+The intuition of the path encoding algorithm is to count the number of paths along “the other” edges and increment by the number of paths. 
+
+<p align="center">
+<img src="https://res.cloudinary.com/dxzx2bxch/image/upload/v1646243822/cs6120/3_dmiy5c.png">
+<p>
+<p align="center">
+<strong> The path encoding generation algorithm </strong>
+</p>
+
+Therefore, we count the number of paths from leaf to node. In other words, we visit vertices in the reverse topological order. 
+
+[Discussion]: based on the discussion in class, we think the time complexity of this algorithm can be represented as both O(n*n) and O(e). Because every edge is only visited once. 
```

Comment:
Also, I'd recommend using bold or something to mark your "discussion" sentences, not square brackets… otherwise, the Markdown parser will think you're trying to make a link. Like this:

```md
**Discussion:** Foo bar…
```